trainers/configurable_evaluator.py

```
@@ -0,0 +1,581 @@
+        return metrics
+
+    def _safe_eval_formula(self, formula: str, variables: Dict[str, Any]) -> float:
+        """Safely evaluate a mathematical formula with restricted operations."""
+        safe_dict = {
+            'metrics': variables.get('metrics', {}),
+            'baseline_metrics': variables.get('baseline_metrics', {}),
+            '__builtins__': {}
+        }
+
+        try:
+            exec(compile(formula, '<string>', 'exec'), safe_dict)
```

Ruff (0.13.1)

Use of `exec` detected

(S102)